### ISSUE TITLE
Union vs. interface types

### ISSUE THREAD

**Issue by nyfarn** [mike-works/typescript-fundamentals#295]
**Is your feature request related to a problem? Please describe.**
If I'm not mistaken, the Basics code example under "Intersection types" actually shows a union type and vice versa.

**Describe the solution you'd like**
The headings could be swapped.

Thank you for this great course.

**Comment by mike-north** [comment]
Good catch! unfortunately I can't change the notes since we've already recorded the course and can't allow the video and code to get out of sync. Definitely something to ensure we get right in the next version!

**Comment by umar-khan** [comment]
Would it be possible to have a note saying that the Intersection and Union notes should be swapped? It might help future students at least until a v3 is up.

I've seen other courses have little helpful popups which show up as blue sections on the video timeline. Eg.

![image](https://user-images.githubusercontent.com/4389490/79594041-58c01380-80aa-11ea-958f-9b4d406733d1.png)

**Comment by hatched-esther** [comment]
Wow.. Im here because I was like... this does not make sense hahah glad to find that Im not crazy.

**Comment by paseaf** [comment]
Actually I find typescript's naming of Union and Intersection types very confusing... I haven't yet found any good explanation of why they are named like that :c

**Comment by nyfarn** [comment]
@paseaf You can think of types as sets of values. The type `boolean` is the set consisting of the values `true` and `false`. The literal type `"foo"` is the set that only contains that string. Now the union type `boolean | "foo"` is exactly the union of those two previous sets, a set that consists of those three values. Similarly, `boolean & "foo"` would be the intersection, the empty set in this case, called `never` in TS.

Now if you talk about how you can *use* those types then you get the opposite result. The properties that you can access on a union type is the intersection of the properties of the individual types. So `boolean | "foo"` has a `toString` method because *both* booleans and strings have it. And the properties that you can access on an intersection type is the union of the properties of the individual types. I'm guess this is what makes these names confusing.

**Comment by paseaf** [comment]
> The properties that you can access on a union type is the intersection of the properties of the individual types.

@nyfarn That's exactly what I was missing! Thank you so much!

I will try to explain it again in case it helps other people
**Definitions**
Let `item` a value of arbitrary type.
Let `Type_A = {foo: number, bar: int}`, `Type_B = {bar: int, baz: string}`

* **the Union Type `Type_A & Type_B`:** 
Defining `item` of `Type_A & Type_B` means `item` is the **combination of** `Type_A` and `Type_B`, which is actually `{foo: number, bar: int, baz: string}`. Then it is safe to allow `item.foo`, `item.bar`, and `item.baz`.
Hence, we have
`Type_A & Type_B = {foo: number, bar: int, baz: string}`

* **the Intersection Type `Type_A | Type_B`:**
Defining `item` of `Type_A | Type_B` means `item` is **either** `Type_A` **or** `Type_B`. Then it is unsafe to allow `item.foo`, because that's an error if the actual `item` is of `Type_B`. Similarly, accessing `item.baz` is also unsafe. 
Hence, we have
`Type_A | Type_B = {bar: int}`

**Comment by paseaf** [comment]
> The properties that you can access on a union type is the intersection of the properties of the individual types.

@nyfarn That's exactly what I was missing! Thank you so much!

After a while of struggling I think I finally understand it.

Here is my understanding, in case it helps other people.
**Definitions**
Let `item` a value of arbitrary type.
Let `Type_A = {foo: number, bar: int}`, `Type_B = {bar: int, baz: string}`

* **the Intersection Type `Type_A & Type_B`:** 
Defining `item` of `Type_A & Type_B` means `item` is the **combination of** `Type_A` and `Type_B`, which is actually `{foo: number, bar: int, baz: string}`. Then it is safe to allow `item.foo`, `item.bar`, and `item.baz`.
Hence, we have
`Type_A | Type_B = {foo: number, bar: int, baz: string}`

* **the Union Type `Type_A | Type_B`:**
Defining `item` of `Type_A | Type_B` means `item` is **either** `Type_A` **or** `Type_B`. Then it is unsafe to allow `item.foo`, because that's an error if the actual `item` is of `Type_B`. Similarly, accessing `item.baz` is also unsafe. 
Hence, we have
`Type_A & Type_B = {bar: int}`

The confusion probably also came from the naming. `Union` means **combining** operation for people who are familiar with set union. While the `Intersection` operator in TS is actually doing set union.
